### ISSUE TITLE
XLA ops parity

### ISSUE THREAD

**Issue by robin5540** [pytorch/xla#2777]
## ❓ Questions and Help
These are probably noob questions:
1. Is there a place where I can find feature parities of torch-xla? A list that shows me if an op or feature is supported on CPU, GPU, and XLA will help a lot. Say, sparse tensors are not supported on TPU.
2. Is there a way to check that my model is XLA compatible without connecting to a Cloud TPU. If not, I think this would be a great feature to have, and I'm willing to help build it. Imagine just hooking all torch APIs and print a warning if something is not TPU compatible.
3. Is there a deep dive design doc about torch-xla? I find it a bit hard to wrap my head around how PyTorch's eager and dynamic tensors are retrofitted for XLA.

**Comment by robin5540** [comment]
a few additional context:
1. I tried to run torch-xla on CPU, but was my python process seems to get stuck at `torch_xla.core.xla_model.get_xla_supported_devices(devkind="CPU")`
2. I'm mainly curious about dataloader. Here is mental disconnect: [GCP documentation](https://cloud.google.com/tpu/docs/tutorials/pytorch-pod) seems to suggest a small VM as the "driver" instance, like a n1-standard-16, would suffice. However, the code seems to require sending CPU data from the "driver" to TPU.

https://github.com/pytorch/xla/blob/7f71a2c81b0323285663fe191b123247482583bd/torch_xla/core/xla_model.py#L837-L846
https://github.com/pytorch/xla/blob/f316e1f3fdd25da5d6c298d03cc0195c7edcdaae/torch_xla/csrc/init_python_bindings.cpp#L753
https://github.com/pytorch/xla/blob/3eaee46ef679cc6a0f1f694bd0a007dbfd09c51b/torch_xla/csrc/tensor_util.cpp#L732

How am I suppose to keep the TPU fed with a n1-standard-16?

Here is an example. Let's say we are training ResNet:
```
image_shape = [N, 3, 900, 600]
image_dtype = fp32

image_size = 4 * 3 * 900 * 600 bytes = 6.18 MB
N = 128
batch_size = 791 MB

batch_time = 500-2000 ms

throughtput_required = 15Gbps - 4Gbps
```

GCP instance can only sustain 7 Gbps egress: https://cloud.google.com/vpc/docs/quota#per_instance
This doesn't seem enough to keep the TPU fed.

**Comment by jysohn23** [comment]
> 1. Is there a place where I can find feature parities of torch-xla? A list that shows me if an op or feature is supported on CPU, GPU, and XLA will help a lot. Say, sparse tensors are not supported on TPU.

We don't have an exhaustive list of documentation, but if you have any specific questions you can feel free to ask. Like you say things like sparse tensors are not supported, or other stuff that needs storage.

> 2. Is there a way to check that my model is XLA compatible without connecting to a Cloud TPU. If not, I think this would be a great feature to have, and I'm willing to help build it. Imagine just hooking all torch APIs and print a warning if something is not TPU compatible.

XLA is a device agnostic linalg compiler, so you can target CPU/GPU/TPU alike. Just set your workers to be XLA:CPU and run your model to then collect a metrics report or run in `PT_XLA_DEBUG=1` mode (if using nightly, will be released as part of 1.8):
```
  export XRT_DEVICE_MAP="CPU:0;/job:localservice/replica:0/task:0/device:XLA_CPU:0"
  export XRT_WORKERS="localservice:0;grpc://localhost:40934"
```

> 3. Is there a deep dive design doc about torch-xla? I find it a bit hard to wrap my head around how PyTorch's eager and dynamic tensors are retrofitted for XLA.

We should be soon releasing a paper on arxiv that should help explain the concept of lazy tensors. In the meantime refer to our API_GUIDE. Our talk at devcon should also be helpful: https://www.youtube.com/watch?v=kdZwgRghy3M

---

The GCP documentation you refer to is using `--fake_data`, which isn't doing any of the CPU heavy JPEG decoding etc, but just sending over a random tensor batch. You're right though once you want to try it on a real dataset and have a heavier input pipeline you may need all the way up to 64+ vCPUs. But note that its 1 GCE VM *per* 8 TPU cores (i.e. 1 TPU host), not 1 GCE VM feeding the entire TPU pod (which would be impossible). But at the end of the day, # vCPU obviously depends on the dataset and model step time.

> GCP instance can only sustain 7 Gbps egress: cloud.google.com/vpc/docs/quota#per_instance

You're referring to `Maximum egress data rate to an external IP address destination`, but GCE VM and your TPU is in the same VPC network and thus communicate through internal IP and limit is much higher. Shouldn't be a bottleneck. In the meantime, our team is working on direct access to TPU VM, which should make a lot of this more usable.
